src/solvers/smt2_incremental/smt2_incremental_decision_procedure.cpp

```diff
@@ -64,14 +65,42 @@ static std::vector<exprt> gather_dependent_expressions(const exprt &expr)
 {
   std::vector<exprt> dependent_expressions;
   expr.visit_pre([&](const exprt &expr_node) {
-    if(can_cast_expr<symbol_exprt>(expr_node))
+    if(
+      can_cast_expr<symbol_exprt>(expr_node) ||
+      can_cast_expr<array_exprt>(expr_node))
     {
       dependent_expressions.push_back(expr_node);
     }
   });
   return dependent_expressions;
 }
 
+void smt2_incremental_decision_proceduret::define_array_function(
+  const array_exprt &array)
+{
+  const auto array_sort =
+    convert_type_to_smt_sort(array.type()).cast<smt_array_sortt>();
+  INVARIANT(
+    array_sort,
+    "Converting array typed expression to SMT should result in a term of array "
+    "sort.");
+  const smt_identifier_termt array_identifier = smt_identifier_termt{
+    "array_" + std::to_string(array_sequence()), *array_sort};
+  solver_process->send(smt_declare_function_commandt{array_identifier, {}});
+  const std::vector<exprt> &elements = array.operands();
+  const std::size_t index_width =
+    array_sort->index_sort().cast<smt_bit_vector_sortt>()->bit_width();
```

Comment:
⛏️ 
`array_sort->index_sort()` is never enforced to be of `smt_bit_vector_sortt` type (although we don't use other types).
I suggest adding and `INVARIANT` to ensure the cast is safe